packages/core-app-api/src/routing/RouteTracker.tsx

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { useEffect, useMemo } from 'react';
+import { matchRoutes, useLocation } from 'react-router-dom';
+import {
+  useAnalytics,
+  AnalyticsContext,
+  CommonAnalyticsContext,
+  RouteRef,
+} from '@backstage/core-plugin-api';
+import { routeObjectCollector } from './collectors';
+import {
+  childDiscoverer,
+  routeElementDiscoverer,
+  traverseElementTree,
+} from '../extensions/traversal';
+import { BackstageRouteObject } from './types';
+
+/**
+ * Returns an extension context given the current pathname and a list of
+ * Backstage route objects.
+ */
+const getExtensionContext = (
+  pathname: string,
+  routes: BackstageRouteObject[],
+): CommonAnalyticsContext | {} => {
+  try {
+    // Find matching routes for the given path name.
+    const matches = matchRoutes(routes, { pathname });
+
+    // Of the matching routes, get the last (e.g. most specific) instance of
+    // the BackstageRouteObject.
+    const routeObject = matches
+      ?.filter(
+        match => (match?.route as BackstageRouteObject).routeRefs?.size > 0,
```

Comment:
Ah, neat.  I always end up creating little filter type guards off to the side--didn't think to inline the guard.

In this case, it looks like not the approach to use, though, since react-router's `matchRoutes()` returns type `RouteMatch[] | null`, and there isn't a built-in way to tell RR that `RouteMatch.route` is of type `BackstageRouteObject` rather than just `RouteObject`.

Though I guess one type cast on the `matchRoutes()` return is better than the two we've got going on here.  Will update for a little more clarity!